### ISSUE TITLE
occasional "zero size buf in output" alert in logs for websocket subscribers

### ISSUE THREAD

**Issue by rmihael** [slact/nchan#132]
Hi. Today we switched our infrastructure to user NChan instead push_stream_module. Everything seems to be running smooth so far, but one there is one alert in logs that was repeated few times over the day:
```
2016/02/22 12:12:08 [alert] 21203#0: *239 zero size buf in output t:0 r:0 f:1 0000000000000000 0000000000000000-0000000000000000 0000000001FAE480 19642-19642, client: 103.236.179.55, server: , request: "GET /push/ws/c533bf989f45db04dea4d36b4/c533bf989f45db04dea4d36b4u53feda21f93a5f34af7229d7?_=1456133205476&tag=&time=&eventid= HTTP/1.1", host: "someserver.com", *239 zero size buf in output t:0 r:0 f:1 0000000000000000 0000000000000000-0000000000000000 0000000001FAE480 19642-19642
```
Path is always the same, as well as the client. There are no 5xx in access logs.

Is it something to be worrying about?

**Comment by slact** [comment]
It may be an indication that a zero-length message was not recognized correctly by Nchan, or (much less likely) a message was lost. I'd like to debug this issue. 

Can you show me your nginx config? Edit out whatever parts you need to keep private. 

What Nchan and Nginx versions are you using? 

What is the output of `nginx -V`?

What kind of subscribers and publishers do you have? How many and how often? Did you send any empty (zero-length) messages?
It looks like this happened to a websocket subscriber. Is that correct?

Have you noticed any missed messages?

**Comment by rmihael** [comment]
Hi. Please see our Nginx config here: https://gist.github.com/rmihael/9d4dfa80729bfe9010bc

We are using Nchan version 0.99.5 and our Nginx is built like this:

```bash
$ nginx -V
nginx version: nginx/1.8.0
built by gcc 4.8.2 20140120 (Red Hat 4.8.2-16) (GCC)
built with OpenSSL 1.0.1k-fips 8 Jan 2015
TLS SNI support enabled
configure arguments: --add-module=../nginx-push-stream-module/ --add-module=../nchan/ --with-http_ssl_module --with-http_gzip_static_module --with-http_stub_status_module --prefix=/usr
```

We only have web sockets subscribers, that's correct. Publish rate is very low, less then 1 message per second on average.

We are logging message sizes that are published and there are no empty messages. There are also no zero-sized POST messages according to Nginx access logs. I think it's safe to conclude that there are no empty messages in our flow.

**Comment by slact** [comment]
> We are logging message sizes that are published and there are no empty messages.

What about received messages? It looks to me like there's some kind of issue when the published message is large enough for nginx to buffer it in a file. You're the first to report this issue, and I'm not one to let a found bug go unfixed, so I'd really like to find out exactly what's going on.

To do that, I'd like you to rebuild nginx so that it crashes and dumps a core when it gets another zero-length buf. 

Would you be willing to do this? I can guide you through the process of rebuilding it and catching the coredumps. It'll take about 30 minutes to set up. Shoot me an email to nchan@slact.net with your contact info and we can set up a convenient time to do this.

**Comment by rmihael** [comment]
I investigated this problem bit deeper and found that these alerts are appearing when we are sending notifications that are larger then usual. I'm talking about sizes of ~25-50 KB. Also that happens not every time when such notifications are sent, most of them are passing fine.
I don't know if users are receiving these messages. I assume they are because we got no complains about it.

One more thing on logs that may be relevant:
```
2016/02/26 10:35:40 [crit] 14484#0: pread() "/mnt/client_body_temp/00624703539" failed (9: Bad file descriptor), pread() "/mnt/client_body_temp/00624703539" failed (9: Bad file descriptor)
```

Speaking of core dump, may be we can start with some extra logging? This problem only manifests itself on production server and I want to avoid Nginx crashes if possible. Could you do a core dump without crashing?

**Comment by slact** [comment]
> Speaking of core dump, may be we can start with some extra logging?

Yes, I'd like to do that as well. 

> Could you do a core dump without crashing?

Yes, that's also possible. E-mail me  at nchan@slact.net and I'll set these things up for you.

**Comment by slact** [comment]
```


for reference:

(gdb) bt full
#0  0x00007f8dc5b40be9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
        resultvar = 0
        pid = 19507
        selftid = 19507
#1  0x00007f8dc5b41fe8 in __GI_abort () at abort.c:90
        save_stage = 2
        act = {__sigaction_handler = {sa_handler = 0x7fff455ccf62, sa_sigaction = 0x7fff455ccf62}, sa_mask = {__val = {140246885257040, 5198025, 122, 4, 
              140734357090240, 47244674368, 56805840, 16384, 0, 0, 0, 21474836480, 140246885256860, 140734357090392, 140246907764736, 140246885269000}}, 
          sa_flags = 5198053, sa_restorer = 0x4f5190 <__PRETTY_FUNCTION__.24276>}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007f8dc5b39d26 in __assert_fail_base (fmt=0x7f8dc5c6fe08 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x4f50e5 "0", 
    file=file@entry=0x4f50c9 "src/core/ngx_output_chain.c", line=line@entry=122, function=function@entry=0x4f5190 <__PRETTY_FUNCTION__.24276> "ngx_output_chain")
    at assert.c:92
        str = 0x3b5fc70 ""
        total = 4096
#3  0x00007f8dc5b39dd2 in __GI___assert_fail (assertion=0x4f50e5 "0", file=0x4f50c9 "src/core/ngx_output_chain.c", line=122, 
    function=0x4f5190 <__PRETTY_FUNCTION__.24276> "ngx_output_chain") at assert.c:101
No locals.
#4  0x000000000040e43f in ngx_output_chain (ctx=0x3ba6008, in=0x0) at src/core/ngx_output_chain.c:122
        bsize = 0
        rc = 0
        last = -2
        cl = 0x3ba60f0
        out = 0x3ba5ff8
        last_out = 0x3ba60f8
        __PRETTY_FUNCTION__ = "ngx_output_chain"
#5  0x000000000045d8f2 in ngx_http_copy_filter (r=0x31a9370, in=0x0) at src/http/ngx_http_copy_filter_module.c:152
        rc = 140734357092592
        c = 0x7f8dc7037b98
        ctx = 0x3ba6008
        clcf = 0x0
        conf = 0x0
#6  0x000000000047beac in ngx_http_range_body_filter (r=0x31a9370, in=0x0) at src/http/modules/ngx_http_range_filter_module.c:599
        ctx = 0x32fa3a0
#7  0x0000000000449172 in ngx_http_output_filter (r=0x31a9370, in=0x0) at src/http/ngx_http_core_module.c:1995
        rc = 57073912
        c = 0x7f8dc7037b98
#8  0x00000000004c41f2 in nchan_output_filter (r=0x31a9370, in=0x0) at ../nchan-0.99.6//src/util/nchan_output.c:183
        clcf = 0x0
        rc = 0
        wev = 0x7f8dbf73a048
        c = 0x7f8dc7037b98
#9  0x00000000004c40e7 in nchan_flush_pending_output (r=0x31a9370) at ../nchan-0.99.6//src/util/nchan_output.c:148
        rc = 0
        wev = 0x7f8dbf73a048
        c = 0x7f8dc7037b98
        clcf = 0x28afa00
#10 0x0000000000455be8 in ngx_http_request_handler (ev=0x7f8dbf73a048) at src/http/ngx_http_request.c:2191
        c = 0x7f8dc7037b98
        r = 0x31a9370
#11 0x0000000000438fee in ngx_epoll_process_events (cycle=0x28951d0, timer=28, flags=1) at src/event/modules/ngx_epoll_module.c:848
        events = 1
        revents = 4
        instance = 1
        i = 0
        level = 4385804
        err = 0
        rev = 0x7f8dbf80b048
        wev = 0x7f8dbf73a048
        queue = 0x7410a0 <ngx_event_timer_sentinel>
        c = 0x7f8dc7037b98
#12 0x000000000042c203 in ngx_process_events_and_timers (cycle=0x28951d0) at src/event/ngx_event.c:248
        flags = 1
        timer = 28
        delta = 1456848247980
#13 0x0000000000436efc in ngx_worker_process_cycle (cycle=0x28951d0, data=0x2) at src/os/unix/ngx_process_cycle.c:767
        worker = 2
        i = 0
        c = 0x407090 <_start>
#14 0x0000000000433df2 in ngx_spawn_process (cycle=0x28951d0, proc=0x436d14 <ngx_worker_process_cycle>, data=0x2, name=0x4f80e0 "worker process", respawn=-3)
    at src/os/unix/ngx_process.c:198
        on = 1
        pid = 0
        s = 2
#15 0x000000000043601a in ngx_start_worker_processes (cycle=0x28951d0, n=4, type=-3) at src/os/unix/ngx_process_cycle.c:357
        i = 2
        ch = {command = 1, pid = 19506, slot = 1, fd = 16}
#16 0x00000000004357dd in ngx_master_process_cycle (cycle=0x28951d0) at src/os/unix/ngx_process_cycle.c:129
        title = 0x28be7fc "master process /usr/sbin/nginx -c /etc/nginx/nginx.conf -p /mnt"
        p = 0x28be83b ""
        size = 64
        i = 5
        n = 140734357091808
        sigio = 42553832
        set = {__val = {0 <repeats 16 times>}}
        itv = {it_interval = {tv_sec = 0, tv_usec = 140734357091632}, it_value = {tv_sec = 15, tv_usec = 19}}
        live = 0
        delay = 42553832
        ls = 0x0
        ccf = 0x2896258
#17 0x0000000000407752 in main (argc=5, argv=0x7fff455cb8f8) at src/core/nginx.c:419
        i = 52
        log = 0x73e9a0 <ngx_log>
        cycle = 0x28951d0
        init_cycle = {conf_ctx = 0x0, pool = 0x2894b90, log = 0x73e9a0 <ngx_log>, new_log = {log_level = 0, file = 0x0, connection = 0, disk_full_time = 0, 
            handler = 0x0, data = 0x0, writer = 0x0, wdata = 0x0, action = 0x0, next = 0x0}, log_use_stderr = 0, files = 0x0, free_connections = 0x0, 
          free_connection_n = 0, reusable_connections_queue = {prev = 0x0, next = 0x0}, listening = {elts = 0x0, nelts = 0, size = 0, nalloc = 0, pool = 0x0}, paths = {
            elts = 0x0, nelts = 0, size = 0, nalloc = 0, pool = 0x0}, open_files = {last = 0x0, part = {elts = 0x0, nelts = 0, next = 0x0}, size = 0, nalloc = 0, 
            pool = 0x0}, shared_memory = {last = 0x0, part = {elts = 0x0, nelts = 0, next = 0x0}, size = 0, nalloc = 0, pool = 0x0}, connection_n = 0, files_n = 0, 
          connections = 0x0, read_events = 0x0, write_events = 0x0, old_cycle = 0x0, conf_file = {len = 21, data = 0x7fff455ccf6b "ss"}, conf_param = {len = 0, 
            data = 0x0}, conf_prefix = {len = 11, data = 0x7fff455ccf6b "ss"}, prefix = {len = 5, data = 0x2894be0 ""}, lock_file = {len = 0, data = 0x0}, hostname = {
            len = 0, data = 0x0}}
        ccf = 0x2896258
(gdb) 
```

**Comment by slact** [comment]
This issue should be now fixed in version 0.99.9

**Comment by slact** [comment]
I'm closing this issue, please reopen if you're still having problems.
